### DISCUSSION TITLE
V2G Control / Bi-directional Charging

### DISCUSSION THREAD

**Discussion by junknik** [lbbrhzn/ocpp#535] [Q&A]
Hi,

Firstly a quick thank you to all of those involved in the project; I just installed the OCPP add-on and it appears that my charger has connected to it :)

After the installation of OCPP I noticed that the amount of data available is quite comprehensive. In particular I was interested by charger.Energy.Active.Export.register and charger.Power.Active>Export, which seemed to indicate that bi-directional chargers may be supported? We have a Quasar bi-directional charger and so this looked promising. It appears as though the current flowing to(from) the charger is set by charger.Maximum_Current. In the Wallbox app for the Quasar this figure can be set between -32A and 32A to allow control of discharging as well as charging. In the OCPP control the values are from 0 to 32A (understandably!) I am not sure but wondered whether modification of this control to allow a negative current to be set would allow me to set discharging on the Quasar. Would it be difficult to modify this?
I am happy to try out any modifications or debug things, but before I started exploring the code I thought I would ask in case such feature was already a WIP.

Thank you!

Neil

**Comment by drc38** [comment]
It'll be interesting to see what is supported with V2X use and the Quasar charger, change the `min_value=0` in `number.py` 

When you have it working we can look at how we include in the codebase, perhaps an if statement based on whether an export measurand is selected on setup.

**Reply by junknik** [reply]
Hi, thanks for the swift response .... I have some good news, and some puzzling news; please forgive me if I am not operating the OCPP add-on correctly I only started playing with it this afternoon!

I changed the `min_value=0 ` in `number.py` to `min_value=-32 ` and restarted HA. I then set the charger.Maximum_Current slider to the minimum value and toggled charger.Charge_Control to initiate the discharge. The good news is that the Quasar started to discharge. The mixed / puzzling result was that it was discharging at a smaller than (I) expected rate, of approx 1.5kW. This is self consistent with what is reported in the charger.Current.Export measurand, which indicates 6.0A.  But I wonder why it is not at (-)32A!

I think that I have seen how to turn logging on so maybe I will take a look shortly and report back with some log entries.

In addition I know that due to the (in this case Chademo) charging protocol but (I beleive it to be the same in a Type 2 AC interface) the current range for charging can be set to between 6A and 32A inclusive.   Discharging with the Quasar has the export range as a mirrored version e.g. -6A to -32A, this means that there will be two gaps (or one gap if current=0 is assumed to be turned off) of 'invalid' values; the Wallbox App just ignores or rounds these values, so perhaps these entries would need to be validated/corrected as well?

**Reply by junknik** [reply]
Please find a screenshot below, I thought it may provide some useful info.
![discharging](https://user-images.githubusercontent.com/106753492/171732853-c33f89b1-0a0a-4c89-a516-8201ee54643a.png)

.

**Reply by drc38** [reply]
See what the logs show, it may be the car is limiting discharge until the battery warms up or similar? On the -6 to 6A other chargers allow values in this range eg 0 to suspend/pause charging.

**Reply by junknik** [reply]
I have been using the Quasar for quite a while through the app and as part of a trial. My experience with it so far is that you can set the discharge rate that you would like through the app and it gets actioned straight away. Will take a look at the logs ......

**Reply by junknik** [reply]
Hi, I don't think that the discharge current is related to a battery temperature or conditioning issue - using the Wallbox app the desired discharge current is reached almost immediately. I turned on a higher level of debugging for the OCPP component, and think that I can see at least part of the issue. When setting a positive current (6A) two messages are sent, the first failing and the second one is accepted thus
```
2022-06-03 15:08:05 DEBUG (MainThread) [custom_components.ocpp] Get Configuration for ChargeProfileMaxStackLevel: 1
2022-06-03 15:08:05 INFO (MainThread) [ocpp] WallboxQuasar: send [2,"430cb035-144c-461a-b56c-52bbd9c26153","SetChargingProfile",{"connectorId":0,"csChargingProfiles":{"chargingProfileId":8,"stackLe
vel":1,"chargingProfileKind":"Relative","chargingProfilePurpose":"ChargePointMaxProfile","chargingSchedule":{"chargingRateUnit":"A","chargingSchedulePeriod":[{"startPeriod":0,"limit":6.0}]}}}]
2022-06-03 15:08:05 INFO (MainThread) [ocpp] WallboxQuasar: receive message [3,"430cb035-144c-461a-b56c-52bbd9c26153",
{
"status": "Rejected"
}]
2022-06-03 15:08:05 DEBUG (MainThread) [custom_components.ocpp] ChargePointMaxProfile is not supported by this charger, trying TxDefaultProfile instead...
2022-06-03 15:08:05 INFO (MainThread) [ocpp] WallboxQuasar: send [2,"c0df942f-288a-409b-bbd1-cae0c53a755a","SetChargingProfile",{"connectorId":0,"csChargingProfiles":{"chargingProfileId":8,"stackLe
vel":0,"chargingProfileKind":"Relative","chargingProfilePurpose":"TxDefaultProfile","chargingSchedule":{"chargingRateUnit":"A","chargingSchedulePeriod":[{"startPeriod":0,"limit":6.0}]}}}]
2022-06-03 15:08:05 INFO (MainThread) [ocpp] WallboxQuasar: receive message [3,"c0df942f-288a-409b-bbd1-cae0c53a755a",
{
"status": "Accepted"
}]
```
When I send a negative current (-6A) they **both** get rejected
```
2022-06-03 15:09:02 INFO (MainThread) [ocpp] WallboxQuasar: send [2,"e8a83593-8a9f-46e3-b230-bd389e687cdb","SetChargingProfile",{"connectorId":0,"csChargingProfiles":{"chargingProfileId":8,"stackLe
vel":1,"chargingProfileKind":"Relative","chargingProfilePurpose":"ChargePointMaxProfile","chargingSchedule":{"chargingRateUnit":"A","chargingSchedulePeriod":[{"startPeriod":0,"limit":-6.0}]}}}]
2022-06-03 15:09:02 INFO (MainThread) [ocpp] WallboxQuasar: receive message [3,"e8a83593-8a9f-46e3-b230-bd389e687cdb",
{
"status": "Rejected"
}]
2022-06-03 15:09:02 DEBUG (MainThread) [custom_components.ocpp] ChargePointMaxProfile is not supported by this charger, trying TxDefaultProfile instead...
2022-06-03 15:09:02 INFO (MainThread) [ocpp] WallboxQuasar: send [2,"55da52d8-ca5c-4370-b321-a1ebe5ed7081","SetChargingProfile",{"connectorId":0,"csChargingProfiles":{"chargingProfileId":8,"stackLe
vel":0,"chargingProfileKind":"Relative","chargingProfilePurpose":"TxDefaultProfile","chargingSchedule":{"chargingRateUnit":"A","chargingSchedulePeriod":[{"startPeriod":0,"limit":-6.0}]}}}]
2022-06-03 15:09:03 INFO (MainThread) [ocpp] WallboxQuasar: receive message [3,"55da52d8-ca5c-4370-b321-a1ebe5ed7081",
{
"status": "Rejected"
}]
2022-06-03 15:09:03 WARNING (MainThread) [custom_components.ocpp] Failed with response: Rejected
```

However it has some affect on the charger as it does start to discharge, albeit at the -6A rate no matter what current limit I specify. I know that the discharge rate can at least be set to -32A, and I believe controlled as the charger is being controlled via OCPP on the trial that I am part of. Perhaps I need to investigate the OCPP spec?

**Reply by drc38** [reply]
Worth seeing if you can "sniff" a successful message from the official app or look at its ocpp documentation.  Otherwise it will be a process of trial and error based on the official ocpp spec.

Also look through the configuration keys in case a custom key is being used for setting the V2X current. It could also be set through a custom data transfer message.

**Reply by drc38** [reply]
You can also try changing the ChargingingProfileKind to TxProfile.

**Reply by junknik** [reply]
Hi, Thanks for the ideas. I don't think that the official app uses OCPP to communicate with the charger; though I guess it is still worth checking. The URL for the OCCP 'server' on the trial starts WSS, so I am assuming that that is a secure (encrypted) link, though I could still take a look. Will also ask about any official documentation!

**Comment by desolina-nz** [comment]
Hi @junknik , just wondering if you got any further with this? I'm about to start on the same journey. I have a Quasar with the latest firmware and about to try to connect to my local HA server over OCPP. Did you fix the issue with discharging past -6A?

**Reply by desolina-nz** [reply]
I have managed to connect through the charger with HA and OCPP. However in the Wallbox app, the OCPP status always remains at "connecting..." so not sure I have it 100% correct.

I can start charging and successfully change the charging current between 6-32A. In numbers.py I change native_min_value to -32. However I cannot get the unit to start discharging when I change Maximum_Current to a negative number, they all seem to get rejected.

I've also tried using the Wallbox HA integration and also am not able to get it into discharge mode. 

Any advice?

**Reply by desolina-nz** [reply]
I am seeing a bunch of warnings "Warning: Set charging profile failed with response Rejected"

**Reply by drc38** [reply]
Have you tried changing the ChargingingProfileKind to TxProfile?

**Reply by desolina-nz** [reply]
no I haven't. How and where would I do that?

**Reply by drc38** [reply]
In `api.py` you can do a find replace

**Reply by desolina-nz** [reply]
oh, ok. So wherever I see `ChargingProfileKind`, just replace with `TxProfile` instead? I see 3 instances in api.py

**Reply by desolina-nz** [reply]
it is always part of `ChargingProfileKindType` so that would be `TxProfileType` ?

**Reply by drc38** [reply]
Try changing:
`ChargingProfilePurposeType.charge_point_max_profile.value`
To
`ChargingProfilePurposeType.tx_profile.value`

**Reply by drc38** [reply]
You can also try changing:
`ChargingProfileKindType.relative.value`
To
`ChargingProfileKindType.absolute.value`

**Reply by desolina-nz** [reply]
Yes that works. Setting a number between 6 and 32 will result in a successful discharge at that rate. If I try to set a negative number, I get the warning "Warning: Set charging profile failed with response Rejected"

**Reply by desolina-nz** [reply]
(I just made the first change you suggested `ChargingProfilePurposeType.tx_profile.value` I have not tried the second change of `relative` to `absolute`)

**Reply by desolina-nz** [reply]
Do I need to put an `if` statement in there to switch between `tx_profile` and `charge_point_max_profile` depending if the inputed max charge rate is negative or positive? Sorry, I'm not much of a software engineer!

**Reply by drc38** [reply]
You can do, are you sure it is as simple as changing the profile to switch between charging and discharging?

**Reply by desolina-nz** [reply]
yeah not really sure! I've got modbus control working pretty well now so might stick to that. Thanks for your help though! Happy to test any changes in the future to the OCPP integration for V2G.
